components/k8s-model-server/http-proxy/server.py

```diff
@@ -38,119 +40,184 @@
+  # 1) if signature_name is specified, try to find it in the signature_map. If
+  # not found, raise an exception.
+  # 2) if signature_name is not specified, check if signature_map only
+  # contains one entry. If so, return the only signature.
+  # 3) Otherwise, use the default signature_name and do 1).
+  if not signature_name and len(signature_map) == 1:
+    return signature_map.keys()[0], signature_map.values()[0]
+
+  key = (signature_name or
+         signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY)
+  if key in signature_map:
+    return key, signature_map[key]
+  else:
+    raise KeyError("No signature found for signature key %s." % signature_name)
 
 class PredictHandler(tornado.web.RequestHandler):
-    @gen.coroutine
-    def post(self, model, version=None):
-        request_key = self.settings['request_key']
-        request_data = tornado.escape.json_decode(self.request.body)
-        instances = request_data.get(request_key)
-        if not instances:
-            self.send_error('Request json object have to use the key: %s'%request_key)
 
-        if len(instances) < 1 or not isinstance(instances, (list, tuple)):
-            self.send_error('Request instances object have to use be a list')
-
-        instances = decode_b64_if_needed(instances)
-
-        input_columns = instances[0].keys()
-
-        request = predict_pb2.PredictRequest()
-        request.model_spec.name = model
-
-        if version is not None:
-            request.model_spec.version = version
-        
-        for input_column in input_columns:
-            values = [instance[input_column] for instance in instances]
-            request.inputs[input_column].CopyFrom(tf.make_tensor_proto(values, shape=[len(values)]))
-
-        stub = self.settings['stub']
-        result = yield fwrap(stub.Predict.future(request, self.settings['rpc_timeout']))
-        output_keys = result.outputs.keys()
-        predictions = zip(*[tf.make_ndarray(result.outputs[output_key]).tolist() for output_key in output_keys])
-        predictions = [dict(zip(*t)) for t in zip(repeat(output_keys), predictions)]
-        self.write(dict(predictions=predictions))
+  @gen.coroutine
+  def post(self, model_name, version_name=None):
+    if not self.settings['signature_map'].get(model_name):
+      self.settings['signature_map'][model_name] = get_signature_map(self.settings['stub'], model_name)
+
+    request_key = self.settings['request_key']
+    request_data = tornado.escape.json_decode(self.request.body)
+    instances = request_data.get(request_key)
+    if not instances:
+      self.send_error('Request json object have to use the key: %s' % request_key)
+    if len(instances) < 1 or not isinstance(instances, (list, tuple)):
+      self.send_error('Request instances object have to use be a list')
+    instances = decode_b64_if_needed(instances)
+    input_columns = instances[0].keys()
+
+    signature_name = request_data.get("signature_name", None)
```

Comment:
`signature_name = request_data.get("signature_name") ` default value of get is already None

Reply:
Done.